### PR TITLE
feat(timings): support dark color scheme in HTML output

### LINKED ISSUE #14586: cargo-timing.html supports dark mode
### Problem

For users using dark mode, opening cargo-timing.html is very glaring

I can use some browser ext to force the web page to switch to dark mode, but the canvas will still be white background.

Image

### Proposed Solution

For web pages, there is already prefers-color-scheme

https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-color-scheme

For canvas, maybe just invert the basic white and black, and keep the other colors?

### Notes

_No response_

### PR DESCRIPTION
**Pull request by ahaoboy**
### What does this PR try to resolve?

Use CSS variables to control colors in different modes

Fixes #14586

### How should we test and review this PR?

Switch to dark mode to preview cargo-timing.html file

### Additional information

I am not an expert in CSS. The color table is generated by gpt. Maybe there are other better color schemes.

Snipaste_2024-09-24_19-36-14
Snipaste_2024-09-24_19-36-50

### CODE REVIEW

File: src/cargo/core/compiler/timings.rs [hunk 1]
```diff
@@ -609,8 +609,59 @@ static HTML_TMPL: &str = r#"
   <title>Cargo Build Timings — {ROOTS}</title>
   <meta charset="utf-8">
 <style type="text/css">
+:root {
+  --error-text: #e80000;
+  --text: #000;
+  --background: #fff;
+  --h1-border-bottom: #c0c0c0;;
```
Comment (weihanglo) [inline src/cargo/core/compiler/timings.rs:616]:
```suggestion
  --h1-border-bottom: #c0c0c0;
```

File: src/cargo/core/compiler/timings.rs [hunk 2]
```diff
@@ -609,8 +609,59 @@ static HTML_TMPL: &str = r#"
...
+    --table-tr-border-top: #333;
+    --table-tr-border-bottom: #333;
+    --table-tr-odd-background: #1e1e1e;
+    --table-td-background: #262626;
+    --table-td-border-right: #333;
+    --canvas-background: #1a1a1a;
+    --axes: #b0b0b0;
+    --grid: #333;
+    --dep-line: #444;
+    --dep-line-highlighted: #fff;
```
Comment (weihanglo) [inline src/cargo/core/compiler/timings.rs:657]: These are all used inside canvas, right? If that is the case I would prefer naming them with a `canvas-` prefix.

File: src/cargo/core/compiler/timings.rs [hunk 3]
```diff
@@ -609,8 +609,59 @@ static HTML_TMPL: &str = r#"
...
+  --grid: #e6e6e6;
+  --block: #aa95e8;
+  --custom-build: #f0b165;
+  --not-custom-build: #95cce8;
+  --dep-line: #ddd ;
+  --dep-line-highlighted: #000;
+  --cpu: rgba(250, 119, 0, 0.2)
+}
+
+@media (prefers-color-scheme: dark) {
```
Comment (weihanglo) [inline src/cargo/core/compiler/timings.rs:638]: Could we split the dark theme support to the other commit?

That is, the first commit would be theming the current CSS color in CSS variables, followed by adding a dark color scheme media qeury.

File: src/cargo/core/compiler/timings.js
Comment (weihanglo) [inline src/cargo/core/compiler/timings.js:1]: I am pretty sure that somebody will complain about dark theme. Not sure if this is a blocker, but should we have a simple theme toggle?

**Review by weihanglo — COMMENTED** [review]
This is cool! Thank you for the contribution.

File: src/cargo/core/compiler/timings.js
Comment (ahaoboy) [reply src/cargo/core/compiler/timings.js:1]: Adding a button to switch modes is more complicated, the button style, position, mode priority handling, etc. Here is an example, if it is acceptable then I will add it to this pr

https://github.com/rust-lang/cargo/commit/b3f4f57464a1da0b249d9db0a3be90159288bf7c

[cargo-timing-toggle.zip](https://github.com/user-attachments/files/17143523/cargo-timing-toggle.zip)

Comment (weihanglo) [reply src/cargo/core/compiler/timings.js:1]: Some colors in the canvas seem broken, but thank you for the attempt :)

I am happy to merge this PR as-is and look forward to your follow-up PR!

**Review by weihanglo — APPROVED** [review]

File: src/cargo/core/compiler/timings.js
Comment (ahaoboy) [reply src/cargo/core/compiler/timings.js:1]: Haha, this example only implements the toggle, missing some css colors, maybe we can add it when someone needs this feature :D

**Pull request merged**